### ISSUE TITLE
Compliance and ATO Prep - Bring Tock up to Code

### ISSUE THREAD

**Issue by ccostino** [18F/tock#652]
Tock is in need of some help in order to be brought up to code, so to speak.  Several compliance-related items must be taken care of to prep for an ATO (and some ought to be done regardless).  These are listed below (and will likely be updated a fair amount), though some may be broken out into their own issues if warranted.

* [ ] Switch to Circle CI from Travis CI (PR #649 is a start, courtesy of @toolness!)
* [ ] Automate deployment with [`autopilot` CloudFoundry plugin](https://github.com/contraband/autopilot) (issue #548 addresses this separately)
* [ ] Make code reviews mandatory for PRs against `master`
* [ ] Update New Relic agent and ensure proper notifications are sent out (create a `tock-dev` Google Group to receive notifications?)
* [ ] Shore up Python dependency installation and management with [`pip-tools`](https://github.com/jazzband/pip-tools/) (including updating all dependencies to latest versions)
* [ ] Introduce and document a structured release process to be tied in with automated deployments (e.g., with [`git flow`](https://github.com/nvie/gitflow) and possibly a tool like [`invoke`](http://www.pyinvoke.org/))
* [ ] Enable automatic app refreshes (not sure if Circle CI supports this as of yet; Travis CI does with `cron` jobs you can schedule)
* [ ] Introduce a tool like [`bandit`](https://wiki.openstack.org/wiki/Security/Projects/Bandit) for static code analysis to help manage security concerns
* [ ] Update Node packages and ensure they run with the latest version of Node and NPM (or [`yarn`](https://yarnpkg.com/en/))
* [ ] Ensure cloud.gov services bound to Tock are set and configured properly
* [ ] Verify logging setup and make adjustments as necessary, especially for any user actions

As mentioned, there might be other things in the future that I could be missing at the moment.  We'll update this as we learn more!

**Comment by abrouilette** [comment]
@hollyallen, FYI.

**Comment by ccostino** [comment]
Switching from Circle CI to Travis CI is underway - I merged @toolness' PR and have another one in the wings to shut off Travis: #654

**Comment by ccostino** [comment]
Just made code reviews mandatory for any merges made to the `master` branch, mirroring FEC's setup (and with an added bonus of ensuring that code is up to date and status checks pass).

**Comment by hollyallen** [comment]
Carlo, please tell me you mean switching to Circle CI. :)

**Comment by ccostino** [comment]
D'oh!  Fixed the statement above about CircleCI vs. Travis. :-)

Thanks for the suggestions on the other tasks, I'll work them into the list!

However, with essentially just myself at a few hours at most each week available for Tock and @annalee whenever she might have a spare minute, I'm concerned about something like the weekly log review.  Who will be able to take that on?

**Comment by ccostino** [comment]
CircleCI switch is done!

**Comment by hollyallen** [comment]
Weekly log review is a strict requirement from GSA OCISO at this time. We can ask to loosen it given that tock is an internal tool that requires a GSA login, I don't want to count on success there. Basically, if someone hacks tock who will discover that?

**Comment by ccostino** [comment]
Oh no, I'm sorry, I'm not debating the importance of the logs or the fact that this is a requirement!  I wouldn't ask for this to change at all (other than identifying what kind of ATO/what level Tock actually needs).  My concern was over who is going to have the actual time to do this review?

I'm going to go over the process with Lindsay for FEC and see what it entails; I'm guessing in this case it'll just have to get worked into the time I have available for Tock (reducing the time I can spend fixing issues, etc.).

In any case, this all speaks to the fact that you and I need to meet soon to actually talk about ATO for Tock, @hollyallen.  I'll schedule something for us soon now that you're back! :-)

**Comment by ccostino** [comment]
Tock was already setup with Gemnasium and alerts are going out just fine, as we received one for Django today.  I double checked the configuration in Gemnasium and all is fine.

**Comment by ccostino** [comment]
The Node.js parts have all been updated to version `8.5.0`, and the New Relic agent has been updated to the latest release as well.

**Comment by ccostino** [comment]
New Relic work is going to be tracked in a separate issue (#667).

**Comment by ccostino** [comment]
New Relic support is now wrapped up.

**Comment by ccostino** [comment]
Just found out from @brittag that Tock is out of compliance purely based on the UAA Auth Route service, which has a known issue (https://github.com/cloudfoundry-community/cf-uaa-guard-service/issues/16).  Opening a new issue separately for tracking this work.

**Comment by abrouilette** [comment]
What’s the need here, expressed as a unit of time?

**Comment by ccostino** [comment]
I'm not sure, but as an estimate I'd say at least 16 - 24 hours (2 - 3 days) minimum of dedicated time for research, implementation, and testing to make sure it's done right and secure.

**Comment by ccostino** [comment]
Furthermore, this likely ties directly into issues #666 and #658, as well as PR #640, so we'd want to make sure these are all addressed as a part of the work.

**Comment by ccostino** [comment]
With #712 merged, that should take care of the two items in the checklist around continuous integration, automated deployments, and a structured and documented release process!  Way to go, @rogeruiz! :-)

**Comment by rogeruiz** [comment]
> Enable automatic app refreshes

@ccostino with this bullet point, is it enough for the application to be `restaged` or does it need to be `redeployed`?

**Comment by ccostino** [comment]
@rogeruiz when we had this working with Travis CI for FEC it was a scheduled job that would build and run through a deployment of the branch associated with a space, e.g., `develop` for our `dev` space and `master` for our `prod` space.  I'm not sure that CircleCI has this capability yet; last I had heard/read it did not, so this may not be possible at the moment.

@afeld, can you shed anymore light on the require for automatic app refreshes as it pertains to ATOs, and do you know if this is something we can do with CircleCI now?

**Comment by rogeruiz** [comment]
@ccostino we can do this with CircleCI and the `workflows.schedule` stanza. But we may need to modify the deployment / release process if we're redeploying. though maybe not. We could always redeploy the latest tag.

If we're refreshing to get the latest updates from :cloud:.gov then I think a regular restage is enough. We shouldn't need to modify the source code of the droplet. https://docs.cloudfoundry.org/devguide/deploy-apps/start-restart-restage.html#restage

> Restaging your application stops your application and restages it, by compiling a new droplet and starting it.

**Comment by ccostino** [comment]
@rogeruiz Yup, that sounds about right, especially with redeploying the latest successful tag.  The only concern I have is trying to maintain zero down-time, hence redeploying with `autopilot` as opposed to just restaging. :-)

**Comment by rogeruiz** [comment]
Seems like a job for [`CF scaleover`](https://github.com/krujos/scaleover-plugin) and a script to watch for the scaling process, if we decide to not redeploy. but yeah you're right. Thanks for clarifying!

**Comment by afeld** [comment]
> can you shed anymore light on the require for automatic app refreshes as it pertains to ATOs

Hmm, not sure that we documented in Before You Ship, but the idea is that short of having intrusion detection and virus scanning in containers in cloud.gov containers, the compromise was to make sure the containers are turned over regularly. This can be done by re-deploying (zero-downtime through scaleover or [autopilot](https://github.com/contraband/autopilot), ideally), [restaging](https://docs.cloudfoundry.org/devguide/deploy-apps/start-restart-restage.html#restage), etc. through the scheduled builds @rogeruiz mentioned.

**Comment by rogeruiz** [comment]
Update Cloud Foundry manifest's to not use depreciated features.

```
Deprecation warning: Specifying app manifest attributes at the top level is deprecated. Found: buildpack, timeout, memory, path, stack.
Please see http://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html#deprecated for alternatives and other app manifest deprecations. This feature will be removed in the future.
```
